### ISSUE TITLE
SAFEARRAY_1D structure

### ISSUE THREAD

**Issue by MarkJohnstoneGitHub** [cristianbuse/VBA-MemoryTools#6]
Is the SAFEARRAY_1D structure correct?
A quick examination it  doesn't apear to match up in structure  size.  
Shouldn't Win64 be pvData As LongPtr instead of LongLong  ?

From 
```
Public Type SAFEARRAY_1D
    cDims As Integer
    fFeatures As Integer
    cbElements As Long
    cLocks As Long
    #If Win64 Then
        dummyPadding As Long
        pvData As LongLong
    #Else
        pvData As Long
    #End If
    rgsabound0 As SAFEARRAYBOUND
End Type
```

**Comment by cristianbuse** [comment]
Hi @MarkJohnstoneGitHub ,

Yes, it is correct. Of course, the structure has only one dimension hence the trailing ```_1D``` in the name but it could have up to 60 dimensions if needed.

Starting with ```VBA7``` support for x64 was introduced. All prior versions starting with VB/VBA6 support **only** x32. On x64 the pointers occupy 8 bytes (or 64 bits) and so a new data type, ```LongLong``` was introduced. To avoid too many conditional compilation blocks of code, Microsoft introduced ```LongPtr``` which is not a true data type. See [here](https://learn.microsoft.com/en-us/office/vba/language/reference/user-interface-help/longptr-data-type) where it is stated:  
> [LongPtr](https://learn.microsoft.com/en-us/office/vba/language/reference/user-interface-help/longptr-data-type) is not a true data type because it transforms to a [Long](https://learn.microsoft.com/en-us/office/vba/language/reference/user-interface-help/long-data-type) in 32-bit environments, or a [LongLong](https://learn.microsoft.com/en-us/office/vba/language/reference/user-interface-help/longlong-data-type) in 64-bit environments. Using LongPtr enables writing portable code that can run in both 32-bit and 64-bit environments. Use LongPtr for pointers and handles.

Quick example. Instead of needing do declare a variable like this:  
```VBA
#If Win64 Then
    Dim x As LongLong
#Else
    Dim x As Long
#End If
```  
you could simply use:  
```VBA
Dim x As LongPtr
```  
which clearly simplifies code. However ```LongPtr``` is not available on previous VB versions (like VB6) and so the ```VBA7``` compiler constant is usually used like so:  
```VBA
#If VBA7 Then
    Dim x As LongPtr
#Else
    Dim x As Long
#End If
```
so that code works on VBA7 and also on prior versions. Again, even in the last example under the VBA7 branch, the ```LongPtr``` could default to ```LongLong``` or ```Long``` depending on bitness. In other words the last block of code actually looks something like this in the compiler's 'eye':  
```VBA
#If VBA7 Then
    #If Win64 Then
        Dim x As LongLong
    #Else
        Dim x As Long
    #End If
#Else
    Dim x As Long
#End If
```

Returning to your question, in the block of code:  
```VBA
#If Win64 Then
    dummyPadding As Long
    pvData As LongLong
#Else
    pvData As Long
#End If
```
there is no need to use ```LongPtr``` because we already know it would default to ```LongLong``` anyway. We, of course, know this because we used the ```Win64``` constant. 

Also, the ```dummyPadding``` must be used because the ```pvData``` pointer must be 8-byte aligned on x64. As you can see there is no need to use the ```VBA7``` compiler constant here as that doesn't tell us the bitness.

In summary, when using ```Win64``` is cleaner to use ```LongLong``` whereas when using ```VBA7``` we usually use ```LongPtr``` to account for both x32 and x64 as well as older VB versions.

**Comment by MarkJohnstoneGitHub** [comment]
Shouldn't it be the X32 requires the padding not the  X64?  

#If Win64 Then
    pvData As LongLong
#Else
    dummyPadding As Long
    pvData As Long
#End If

Too early think if the padding is in the correct  order.  Makes me think if I need to fix my X32 SafeArray  structures.

I prefer using the LongPtr when referring to addresses.   Bit rusty on  SafeArrays was organising my old SafeArray and memory functions was checking out your memory library.  Not sure  you had a FillMemory function.

I did some safearray manipulation to  process dictionary keys.
https://github.com/MarkJohnstoneGitHub/VBA-IDictionary/blob/master/scr/ManagedCharSafeArray.cls

https://github.com/MarkJohnstoneGitHub/VBA-IDictionary/blob/master/archive/Archive%20V2.0/src/TextEncoderUnicode.cls

When get around it do a refresher on SafeArrays.  Not sure in the memory code if had a memory dump/display routine.  I should have one somewhere.  Might take a bit of digging around.

**Comment by cristianbuse** [comment]
The x32 structure is 4-byte aligned so no padding needed. Definitely needed for x64.

When I have some time I will look into your dictionary repo.

Also, I have no FillMemory method but maybe I will add one if useful.

**Comment by MarkJohnstoneGitHub** [comment]
Argh yes you're  correct the X64 requires the padding.  
I checked out my old reply on https://stackoverflow.com/questions/18784470/where-is-safearray-var-type-stored/57729377#57729377 

The memory dump  I posted displays the four bytes of padding required. 

I think saw some old code I had in the TextEncoderUnicode.cls with fill memory.  Found a some memory dump code I used for testing possible needs some cleaning up.      

Wasn't planning on revisting SafeArrays.   Looks I need  to  brush on  SafeArrays and while at a decent memory dump utility. Some nice memory dump functions could come in handy.
